src/main.rs

```diff
@@ -107,78 +122,100 @@ fn mem_parser(arg: &str) -> Result<usize, String> {
   Ok(base * mult)
 }
 
-impl WOpts {
-  fn get_warning_opts(self) -> WarningOpts {
-    let mut warning_opts = WarningOpts::default();
-    let argm = Args::command().get_matches();
-    if let Some(wopts_id_seq) = argm.get_many::<clap::Id>("WOpts") {
-      let allows = &mut self.allows.into_iter();
-      let denies = &mut self.denies.into_iter();
-      let warns = &mut self.warns.into_iter();
-      WarningOpts::from_cli_opts(&mut warning_opts, wopts_id_seq.collect(), allows, denies, warns);
-    }
-    warning_opts
-  }
-}
-
 fn main() {
   fn run() -> Result<(), String> {
     #[cfg(not(feature = "cli"))]
     compile_error!("The 'cli' feature is needed for the hvm-lang cli");
 
-    let args = Args::parse();
-    let mut opts = Opts::light();
-    Opts::from_vec(&mut opts, args.opts)?;
-    let warning_opts = args.wopts.get_warning_opts();
-
-    let mut book = load_file_to_book(&args.path)?;
-    if args.verbose {
-      println!("{book:?}");
-    }
+    let cli = Cli::parse();
+    let opts = Opts::light();
+    let arg_verbose = cli.verbose;
 
-    match args.mode {
-      Mode::Check => check_book(book)?,
-      Mode::Compile => {
-        let compiled = compile_book(&mut book, opts)?;
+    let verbose = |book: &_| {
+      if arg_verbose {
+        println!("{book:?}");
+      }
+    };
 
-        for warn in &compiled.warnings {
-          eprintln!("WARNING: {warn}");
-        }
+    run_mode(cli, &verbose, opts)?;
 
-        print!("{}", show_book(&compiled.core_book));
-      }
-      Mode::Desugar => {
-        desugar_book(&mut book, opts)?;
-        println!("{book}");
-      }
-      Mode::Run => {
-        let mem_size = args.mem / std::mem::size_of::<(hvmc::run::APtr, hvmc::run::APtr)>();
-        let (res_term, def_names, info) =
-          run_book(book, mem_size, !args.single_core, args.debug, args.linear, warning_opts, opts)?;
-        let RunInfo { stats, readback_errors, net: lnet } = info;
-        let total_rewrites = total_rewrites(&stats.rewrites) as f64;
-        let rps = total_rewrites / stats.run_time / 1_000_000.0;
-        if args.verbose {
-          println!("\n{}", show_net(&lnet));
-        }
-
-        println!("{}{}", readback_errors.display(&def_names), res_term.display(&def_names));
-
-        if args.stats {
-          println!("\nRWTS   : {}", total_rewrites);
-          println!("- ANNI : {}", stats.rewrites.anni);
-          println!("- COMM : {}", stats.rewrites.comm);
-          println!("- ERAS : {}", stats.rewrites.eras);
-          println!("- DREF : {}", stats.rewrites.dref);
-          println!("- OPER : {}", stats.rewrites.oper);
-          println!("TIME   : {:.3} s", stats.run_time);
-          println!("RPS    : {:.3} m", rps);
-        }
-      }
-    }
     Ok(())
   }
   if let Err(e) = run() {
     eprintln!("{e}");
   }
 }
+
+fn run_mode(cli: Cli, verbose: &dyn Fn(&hvml::term::Book), mut opts: Opts) -> Result<(), String> {
```

Comment:
This name is a bit strange because we have a mode called "run".
Maybe something like `execute_cli_mode` to avoid the confusion.